src/apps/basic/basic_apps.lua

```diff
@@ -6,11 +6,40 @@ local app = require("core.app")
 local freelist = require("core.freelist")
 local packet = require("core.packet")
 local link = require("core.link")
+local ffi = require("ffi")
 local transmit, receive = link.transmit, link.receive
 
+--- # `Truncate` app: truncate or zero pad packet to length n
+Truncate = {}
+function Truncate:new(n)
+   return setmetatable({n = n}, {__index=Truncate})
+end
+function Truncate:push()
+   for _ = 1, link.nreadable(self.input.input) do
+      local p = receive(self.input.input)
+      ffi.fill(p.data, math.min(0, self.n - p.length))
```

Comment:
I think documenting the ABI is a separate issue, since that would not define the FFI field names (`data`/`length`) if I understand correctly(?).

One advantage (imho) of the API functions is that they are read-only. I am fine with documenting the field names as well and encouraging their use, or even getting rid of the API functions. Either way, I would like that API to be defined before I recommend its use.

I have encountered direct uses of the fields before as well, but considered them as semi-bugs until now.